projects/vdk-plugins/vdk-notebook/src/vdk/plugin/notebook/notebook.py

```diff
@@ -0,0 +1,120 @@
+# Copyright 2021 VMware, Inc.
+# SPDX-License-Identifier: Apache-2.0
+import json
+import logging
+import pathlib
+from dataclasses import dataclass
+from pathlib import Path
+from typing import List
+
+from vdk.internal.builtin_plugins.run.file_based_step import TYPE_PYTHON
+from vdk.internal.builtin_plugins.run.file_based_step import TYPE_SQL
+from vdk.internal.builtin_plugins.run.job_context import JobContext
+from vdk.internal.core import errors
+from vdk.plugin.notebook.notebook_based_step import NotebookStep
+from vdk.plugin.notebook.notebook_based_step import NotebookStepFuncFactory
+
+log = logging.getLogger(__name__)
+
+
+class JobNotebookLocator:
+    """
+    Locate the data job files that would be executed by us.
+    """
+
+    @staticmethod
+    def get_notebook_files(directory: pathlib.Path) -> List[pathlib.Path]:
+        """Locates the files in a directory, that are supported for execution.
+        Files supported for execution are: .ipynb
+        Other files in the directory are ignored.
+        :return: List of files from the directory that supported for execution, sorted alphabetically by name.
+        :rtype: :class:`.list`
+        """
+        script_files = [
+            x for x in directory.iterdir() if (x.name.lower().endswith(".ipynb"))
+        ]
+        script_files.sort(key=lambda x: x.name)
+        log.debug(f"Script files of {directory} are {script_files}")
+        return script_files
+
+
+class Notebook:
+    """
+    Given a notebook file locates the cells with "vdk" tag and saves them.
+    Files supported for reading are: ipynb.
+    Other cells are ignored.
+    """
+
+    def __init__(self, file_path: Path):
+        try:
+            self.sql_and_run_cells = []
+            self.python_helper_cells = []
+            self.file_path = file_path
+
+            content = json.loads(file_path.read_text())
+            cells = [
+                cell
+                for cell in content["cells"]
+                if cell["cell_type"] == "code"
+                and "vdk" in cell["metadata"].get("tags", {})
+            ]
+            for cell in cells:
+                code = "".join(cell["source"])
+                if code.startswith("%sql") or "def run(" in code:
+                    if code.startswith("%sql"):
+                        code = "".join(cell["source"])
+                        code.replace(";", "")
+                        self.sql_and_run_cells.append(
+                            [code.replace("%sql", ""), TYPE_SQL]
+                        )
+                    else:
+                        self.sql_and_run_cells.append(
+                            ["".join(cell["source"]), TYPE_PYTHON]
+                        )
+                else:
+                    self.python_helper_cells.append("".join(cell["source"]))
```

Comment:
btw, I really like predicates, because they explain the design better due naming them, and could be reused.
For example a named lambda(/function) `code.startswith("%sql")` applied to code: str, what would be....`is_code_sql`
and `"def run(" in code` could be `code_is_python`

Then you may consider moving those top-level, so within method implementation you reuse them, and a `notebook.py` has this functionallity to distinguish between code that is sql or python. And in case you need to change the implementation, you may do within the lambdas(/functions) themselves, or encapsulated.